File: python/utils/validateAndConvertIntFields.py
import os
import numpy as np
from scipy.io import loadmat

def validate_and_convert_int_fields(file, fields):
    """
    Load a MATLAB .mat file, validate that specified fields contain integer values,
    and convert those fields to numpy.int32.

    The function expects the input file to contain a MATLAB struct accessible
    via the loaded object. It will extract that struct into a Python dict of
    fields and then validate/convert each requested field.

    Parameters
    ----------
    file : str
        Path to a .mat file to load with scipy.io.loadmat.
    fields : list of str
        Names of fields within the loaded struct to validate and convert.

    Returns
    -------
    dict
        Dictionary of fields from the loaded struct, with the specified fields
        converted to numpy.int32.

    Raises
    ------
    AssertionError
        If the provided path is not a file or the loaded object does not expose
        a MATLAB struct via __dict__.
    KeyError
        If any requested field is missing from the struct.
    TypeError
        If a requested field is not numeric.
    ValueError
        If a requested field contains non-integer numeric values.

    Example
    -------
    s = validate_and_convert_int_fields('subject.mat', ['ecog_chan_idx', 'depth_chan_idx'])
    events = validate_and_convert_int_fields('events.mat', ['start_idx', 'end_idx', 'chan_idx'])
    """
    assert os.path.isfile(file), f"Input must be a valid file path: {file}"

    # Load the .mat file. We use squeeze_me and struct_as_record to get simpler arrays/objects.
    struct = loadmat(file, squeeze_me=True, struct_as_record=False)

    for f in fields:
        if f not in struct:
            raise KeyError(f'Field "{f}" not found in input dictionary.')

        # Ensure we operate on a numpy array for dtype checks and casting.
        arr = np.array(struct[f])
        # Require numeric dtype (int/float)
        if not np.issubdtype(arr.dtype, np.number):
            raise TypeError(f'Field "{f}" must be numeric.')
        # Ensure all values are integers (no fractional part)
        if not np.all(np.equal(np.mod(arr, 1), 0)):
            raise ValueError(f'Field "{f}" contains non-integer values.')

        # Convert to 32-bit integers for consistency.
        struct[f] = arr.astype(np.int32)

    return struct

File: matlab/utils/validateAndConvertIntFields.m
function s = validateAndConvertIntFields( struct_file, fields )
%VALIDATEANDCONVERTINTFIELDS Validate specified struct fields are integers and convert to int32.
%
% Checks that the specified fields in struct contain only integer
% values, errors if not, then converts those fields to int32.
%
% Inputs:
%   struct_file - string, filepath of struct
%   fields      - cell array of strings, names of fields to check & convert
%
% Output:
%   s           - output struct with specified fields converted to int32
%
% Example usage:
%   channel_info = validateAndConvertIntFields( channel_info_file,  {'ecog_chan_idx', 'depth_chan_idx'});
%   events = validateAndConvertIntFields(events, {'start_idx', 'end_idx', 'chan_idx'});
%
% Written by Shi-Bei (Ashley) Tan (tashibei@umich.edu) in 2025
% Copyright © 2025 by Shi-Bei (Ashley) Tan is licensed under CC BY-NC 4.0

assert(isfile(struct_file), 'File not found: %s', struct_file);
s = load(struct_file);

for i = 1:numel( fields )
    f = fields{i};
    if ~isfield( s, f )
        error( 'Field "%s" not found in input struct.', f );
    end
end

end